psi4/src/psi4/libfock/composite.cc

```diff
@@ -0,0 +1,796 @@
+#include "composite.h"
+#include "jk.h"
+
+#include "psi4/libmints/integral.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/lib3index/dftensor.h"
+#include "psi4/libqt/qt.h"
+#include "psi4/libpsi4util/PsiOutStream.h"
+
+#include <vector>
+#include <unordered_set>
+#include <algorithm>
+#include <sstream>
+
+#ifdef _OPENMP
+#include <omp.h>
+#include "psi4/libpsi4util/process.h"
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+SplitJKBase::SplitJKBase(std::shared_ptr<BasisSet> primary, Options& options) : primary_(primary), options_(options) {
+
+    print_ = options_.get_int("PRINT");
+    debug_ = options_.get_int("DEBUG");
+    bench_ = options_.get_int("BENCH");
+
+    nthread_ = 1;
+#ifdef _OPENMP
+    nthread_ = Process::environment.get_n_threads();
+#endif
+
+}
+
+CompositeJK::CompositeJK(std::shared_ptr<BasisSet> primary, std::shared_ptr<BasisSet> auxiliary, 
+                         std::string jtype, std::string ktype, Options& options)
+                        : JK(primary), auxiliary_(auxiliary), jtype_(jtype), ktype_(ktype), options_(options) { common_init(); }
+
+void CompositeJK::common_init() {
+
+    nthread_ = 1;
+#ifdef _OPENMP
+    nthread_ = Process::environment.get_n_threads();
+#endif
+
+    if (jtype_ == "DIRECT_DF") {
+        jalgo_ = std::make_shared<DirectDFJ>(primary_, auxiliary_, options_);
+    } else {
+        throw PSIEXCEPTION("J BUILD TYPE " + jtype_ + " IS NOT SUPPORTED IN COMPOSITE JK!");
+    }
+
+    if (ktype_ == "LINK") {
+        kalgo_ = std::make_shared<LinK>(primary_, options_);
+    } else {
+        throw PSIEXCEPTION("K BUILD TYPE " + ktype_ + " IS NOT SUPPORTED IN COMPOSITE JK!");
+    }
+}
+
+size_t CompositeJK::memory_estimate() {
+    return 0;   // only Direct-based integral algorithms are currently implemented in Composite JK
+}
+
+void CompositeJK::preiterations() {}
+void CompositeJK::postiterations() {}
```

Comment:
If preiterations() and postiterations() are empty, it might be better to put "return;" in them to make it more clear that they are essentially empty functiopns doing nothing.